processing/src/main/java/org/apache/druid/frame/key/ClusterByPartitions.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.key;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import org.apache.druid.java.util.common.IAE;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Holder object for a set of {@link ClusterByPartition}. There are no preconditions put upon the partitions, except
+ * that there is at least one of them.
+ *
+ * In particular, they are not required to abut each other. See {@link #allAbutting()} to check if this particular list
```

Comment:
While the partitions may not be abutting, is there any implicit restriction on overlapping partitions?

Reply:
The class doesn't enforce anything, including anything related overlappingness. `allAbutting` can be used to check that the partitions are all abutting and non-overlapping (because abutting implies non-overlapping). I added some detail to the comment to hopefully clarify this.